src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/CheckForTypeRename.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System.ComponentModel.Composition;
+using System.Globalization;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Rename
+{
+    [Export(typeof(IProjectTreeActionHandler))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    [Order(10)]
+    internal class CheckForTypeToRename : ProjectTreeActionHandlerBase
+    {
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IEnvironmentOptions _environmentOptions;
+        private readonly IUserNotificationServices _userNotificationServices;
+        private readonly IRenameTypeService _renameTypeService;
+
+        [ImportingConstructor]
+        public CheckForTypeToRename(IUnconfiguredProjectVsServices projectVsServices,
+                                  IEnvironmentOptions environmentOptions,
+                                  IUserNotificationServices userNotificationServices,
+                                  IRenameTypeService renameTypeService)
+        {
+            _projectVsServices = projectVsServices;
+            _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
+            _renameTypeService = renameTypeService;
+        }
+
+        public override async Task RenameAsync(IProjectTreeActionHandlerContext context, IProjectTree node, string value)
+        {
+            if (!node.Flags.Contains(ProjectTreeFlags.SourceFile))
+            {
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            string filePath = node.FilePath;
+
+            // Check if there are any symbols that need to be renamed
+            if (!await _renameTypeService.DoesFileNameMatchTypeAsync(filePath))
+            {
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            string name = Path.GetFileNameWithoutExtension(filePath);
+
+            // Ask if the user wants to rename the symbol
+            bool userConfirmed = await CheckUserConfirmationAsync(name);
+            if (!userConfirmed)
+            {
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            // queue symbol to be renamed
+            _renameTypeService.QueueFileWithTypeToBeRenamed(filePath);
+
+            await base.RenameAsync(context, node, value);
+        }
+
+        private async Task<bool> CheckUserConfirmationAsync(string oldFileName)
+        {
```

Comment:
Side note, also you've said it is out of the scope of the PR, I think this actually addresses https://github.com/dotnet/project-system/issues/1020 because the prompt is now showed at the correct location; as the file is being renamed.